groupbalancer.go

```diff
@@ -185,3 +365,56 @@ func findGroupBalancer(protocolName string, balancers []GroupBalancer) (GroupBal
 	}
 	return nil, false
 }
+
+// findRack is the default rack resolver strategy.  It currently only supports
+// Linux EC2 deployments and will return the name of the availability zone.
+func findRack() string {
+	switch whereAmI() {
+	case "aws":
+		return awsAvailabilityZone()
+	}
+	return ""
+}
+
+func whereAmI() string {
+	// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html
+	for _, path := range [...]string{
+		"/sys/devices/virtual/dmi/id/product_uuid",
+		"/sys/hypervisor/uuid",
+	} {
+		b, err := ioutil.ReadFile(path)
+		if err != nil {
+			continue
+		}
+		s := string(b)
+		switch {
+		case strings.HasPrefix(s, "EC2"), strings.HasPrefix(s, "ec2"):
+			return "aws"
+		}
+	}
+	return "somewhere"
+}
+
+// awsAvailabilityZone queries the metadata endpoint to discover the
+// availability zone where this code is running.  we avoid calling this function
+// unless we know we're in AWS.  Otherwise, in other environments, we would need
+// to wait for the request to 169.254.169.254 to timeout before proceeding.
+func awsAvailabilityZone() string {
```

Comment:
FWIW : While on aws this may not work for teams running on ecs and blocking the ec2 metadata.

Reply:
Good call.  I also found that there's a better way to do this on ECS, so I pushed up some new code.  Now it will try to use the local ECS metadata and will fail back to trying to EC2 if that doesn't work out.

Reply:
LGTM !